### ISSUE TITLE
Consider interface to filling data into already allocated arrays

### ISSUE THREAD

**Issue by botev** [mhiley/rust-netcdf#11]
I'm considering using this library for Machine Learning purposes, as I need to speed up loading random samples from a large netcdf file, which in python is prohibitively slow. However, for the sake of speed, it would be nice if the `array_at`, `values_at` and all of the other interfaces have also methods to which you can provide an already preallocated memory array and it will fill in the values there, rather than allocating a new one.

**Comment by bluss** [comment]
I'm watching for this feature as well. I have been using an untyped (`*const u8`) interface and would be interested in that kind of low level interface as well, or enough low level access that it can be implemented using the right variable ids with the netcdf-sys interface.

**Comment by evomassiny** [comment]
Hi, 
I've implemented an alternative to `values_at` and `values` that can be used with a pre-allocated vector, see the [related PR](https://github.com/mhiley/rust-netcdf/pull/12).
As for the equivalent of `array_at`, I don't know how to mutate the raw data of an ArrayD (yet).

@bluss The netcdf-sys interface expose all the "nc_get_vara_*" fonction familly from the [original C library](https://www.unidata.ucar.edu/software/netcdf/netcdf-4/newdocs/netcdf-c/nc_005fget_005fvara_005f-type.html), Is this what you need ?

**Comment by bluss** [comment]
Sure, I'm thinking of `nc_get_vara` which uses a `void *`. I have a prototype (using netcdf-sys) that is using it:

```rust
    /// Read variable data of any type; without conversion
    ///
    /// # Safety
    ///
    /// The caller must ensure that `ptr` points to a place where the output can
    /// be written safely.
    pub unsafe fn read_at(&self, indices: &[size_type], counts: &[size_type], ptr: *mut u8) -> Result<(), Ety>
    {
        netcdf_lock!();
        check(ns::nc_get_vara(self.group_id, self.id, indices.as_ptr(), counts.as_ptr(), ptr as _))
    }
```

I have been struggling to find the right way to work with a "foreign typesystem" (netCDF types in this case) and sometimes using enums everywhere seems both redundant and restrictive. The C approach to "generic" is sometimes really effective.

**Comment by mhiley** [comment]
closed by #21
